### PR TITLE
THREESCALE-9308 HeadCrab Vulnerability (Redis)

### PR DESCRIPTION
**Pull request by valerymo**
Jira:  https://issues.redhat.com/browse/THREESCALE-9308

**Will Open new PR and close this one as significant changes in master, including GO version**  
**Please continue review in PR https://github.com/3scale/3scale-operator/pull/954**   
**All tests will be repeated in new PR - 954**

PR is intended to prevent redis replication, slaves creation and revert any change made to the Redis configuration - `redis-config` config map 
- for Internal Redis Only.

- If Redis is Internal than following records are added to configuration:
rename-command REPLICAOF ""
rename-command SLAVEOF ""

- A controller will revert any changes made to the Redis configuration - `redis-config` config map.

## Tests Preparation - files
- s3-creds-secret.yaml

```
kind: Secret
apiVersion: v1
metadata: 
  name: s3-credentials
  namespace: 3scale-test
data: 
  AWS_ACCESS_KEY_ID: QUtJQVY2SVcccc
  AWS_SECRET_ACCESS_KEY: aU5VbWdZY3hjSDFccc
  AWS_BUCKET: dm1vY2NzZjZxOGxyZWRoYXRyaGccccc
  AWS_REGION: ZXUtd2Vcccc
type: Opaque
```

- apimanagerCR.yaml

```
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: example-apimanager
spec:
  system: 
    fileStorage: 
      simpleStorageService: 
        configurationSecretRef: 
          name: s3-credentials
  wildcardDomain: apps.xxxxxx.axxxx.xxx.xxxxx.org

```

## Validation

-  get this PR  / checkout branch `THREESCALE-9308`

cd .../3scale-operator and run 3scale installation:

```
make install
oc new-project 3scale-test
make download
make run
```

- in another xterm:

```
oc apply -f s3-creds-secret.yaml
oc apply -f apimanagerCR.yaml
```

- Check `redis-config` ConfigMap has required changes:
```
(THREESCALE-9308)$ oc describe cm redis-config |grep rename-command
rename-command REPLICAOF ""
rename-command SLAVEOF ""
```

- Check that replicaof and slaveof commands are disabled.
Expected behavior as in example below

```
kubectl exec -ti $(kubectl get pods -l deployment=backend-redis -o name) -- /bin/bash
bash-4.2$ redis-cli 
127.0.0.1:6379> replicaof 10.10.10.10 6379
(error) ERR unknown command `replicaof`, with args beginning with: `10.10.10.10`, `6379`,
127.0.0.1:6379> replicaof
(error) ERR unknown command `replicaof`, with args beginning with: 
127.0.0.1:6379> slaveof
(error) ERR unknown command `slaveof`, with args beginning with: 
127.0.0.1:6379> 
```

Notes: Internally, within redis system, the `slaveof` and `replicaof` commands will be disabled. 
So the only way to enable it would be changing the configuration 
See next item  - that operator is controllilng `redis-config` config map and  will revert any change made to the configuration.

- Check that `redis-config` config map changes are reverted
    - Edit config map - delete `rename-command REPLICAOF ""` and `rename-command SLAVEOF ""`
    ```
    $ oc edit cm redis-config
    ......
    ......
    // delete these lines
      rename-command REPLICAOF ""
      rename-command SLAVEOF ""
    ```
    ```
    $ oc edit cm redis-config
    configmap/redis-config edited
    ```

    - Check configMap and see that these lines are exists (reconsiler reverted configmap) 
    ```
    rename-command REPLICAOF ""
    rename-command SLAVEOF ""
    .....
    Events:  <none>
    $ oc describe cm redis-config
    ```
### Check Upgrade
**Following validation was done**
- Create Images/Indexes for master and for branch [THREESCALE-9308](https://issues.redhat.com//browse/THREESCALE-9308)
-  Install master ; check that redis-config CM does Not contain rename-command for REPLICAOF and SLAVEOF 
```
$ oc describe cm redis-config |grep rename-com
```
- Do upgrade  and check that Redis configuration was updated, new records added: 
```
$ oc describe cm redis-config |grep rename-com
rename-command REPLICAOF ""
rename-command SLAVEOF ""
```

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -5,6 +5,26 @@ metadata:
   creationTimestamp: null
   name: manager-role
 rules:
+- apiGroups:
```
Comment (eguzki) [inline config/rbac/role.yaml:8]: duplicated. This `configmaps` resource has already been assigned to permissions.

File: main.go [hunk 2]
```diff
@@ -363,6 +363,18 @@ func main() {
 		setupLog.Error(err, "unable to create controller", "controller", "Application")
 		os.Exit(1)
 	}
+
+	if err = (&capabilitiescontroller.RedisConfigCMReconciler{
```
Comment (eguzki) [inline main.go:367]: There is no need for a new controller. 

I propose a new approach: Use the existing apimanager_controller.  You will need to add new watcher for owned (with ownerrefs) configmaps. Maybe filtering only for redis configmaps using labels. Then change the mutator, which currently is "CreateOnlyMutator", to something more restrictive. 

If you implement a new controller, the apimanager controller creates the config map and another controller ensures nobody change it. Maybe better to keep all the logic in one controller.

File: controllers/capabilities/redis_threescale_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	"reflect"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type RedisConfigCMReconciler struct {
+	*reconcilers.BaseReconciler
+	previousConfigMap *corev1.ConfigMap
```
Comment (eguzki) [inline controllers/capabilities/redis_threescale_reconciler.go:18]: usually this approach is not recommended. The state of a controller needs to be read from the cluster... cannot rely on local status. The rationale being what if the operator controller is being restarted?

**Review by eguzki — COMMENTED** [review]
Not requesting for changes. Just giving how I would have done this. 

Up to you the approach you want to follow. 

Something to look at: upgrade. Let's say there is 2.14 installed and a user installs a new operator with this codebase... the configmap will be changed. Redis is being rolled out? automatically or the operator needs to do something else?

File: controllers/capabilities/redis_threescale_reconciler.go (hunk 3 above)
Comment (valerymo) [reply controllers/capabilities/redis_threescale_reconciler.go:18]: but I will continue with your suggestion wihout new controller. Thank you

File: main.go (hunk 2 above)
Comment (valerymo) [reply main.go]: Done. Thank you

File: config/rbac/role.yaml (hunk 1 above)
Comment (valerymo) [reply config/rbac/role.yaml]: done. Thank you

File: controllers/capabilities/redis_threescale_reconciler.go (hunk 3 above)
Comment (valerymo) [reply controllers/capabilities/redis_threescale_reconciler.go]: done. Thank you

**Comment by valerymo** [comment]
> Not requesting for changes. Just giving how I would have done this.
> 
> Up to you the approach you want to follow.
> 
> Something to look at: upgrade. Let's say there is 2.14 installed and a user installs a new operator with this codebase... the configmap will be changed. Redis is being rolled out? automatically or the operator needs to do something else?

- Approach - used as @eguzki  suggested. Done & tested
- Regarding Upgrade - I'm not completely sure yet, if new CM will be installed (?) If not - seems need add code that checking if replicaof and slaveof records exist in config (?)

**Comment by valerymo** [comment]
> > Not requesting for changes. Just giving how I would have done this.
> > Up to you the approach you want to follow.
> > Something to look at: upgrade. Let's say there is 2.14 installed and a user installs a new operator with this codebase... the configmap will be changed. Redis is being rolled out? automatically or the operator needs to do something else?
> 
> * Approach - used as @eguzki  suggested. Done & tested
> * Regarding Upgrade - I'm not completely sure yet, if new CM will be installed (?) If not - seems need add code that checking if replicaof and slaveof records exist in config (?)

3scale  operator Upgrade - checked. `redis-config` ConfigMap was updated to new one that contains records`rename-command REPLICAOF ""` and `rename-command SLAVEOF ""`

**Comment by valerymo** [comment]
Evidencese that I can provide now (next day after Upgrade done):

Screenshot from 2024-03-13 07-53-00

// Time of CatalogSource update - Upgrade started
9945  2024-03-12 16:02:23 oc edit catalogsource threescale-operators --namespace=openshift-marketplace
//Time now
$ date
Wed Mar 13 07:55:19 IST 2024

**Comment by eguzki** [comment]
> > > Not requesting for changes. Just giving how I would have done this.
> > > Up to you the approach you want to follow.
> > > Something to look at: upgrade. Let's say there is 2.14 installed and a user installs a new operator with this codebase... the configmap will be changed. Redis is being rolled out? automatically or the operator needs to do something else?
> > 
> > 
> > 
> > * Approach - used as @eguzki  suggested. Done & tested
> > * Regarding Upgrade - I'm not completely sure yet, if new CM will be installed (?) If not - seems need add code that checking if replicaof and slaveof records exist in config (?)
> 
> > 3scale  operator Upgrade - checked. `redis-config` ConfigMap was updated to new one that contains records`rename-command REPLICAOF ""` and `rename-command SLAVEOF ""`

Yes, we know that. Redis is being rolled out? automatically or the operator needs to do something else? Is the redis data still available? I expect so, because it relies on a PVC, but good to test that as well.

**Comment by eguzki** [comment]
> @eguzki, could you please tell me if this checking is looks sufficient or maybe we need check more. Thank you for review

Looks like redis is **not** being rolled out, hence the configmap change might not be applied. Can you verify that? after the upgrade, you can check if the command is available:
```
kubectl exec -ti $(kubectl get pods -l deploymentconfig=backend-redis -o name) -- /bin/bash
bash-4.2$ redis-cli 
127.0.0.1:6379> replicaof 10.10.10.10 6379
(error) ERR unknown command `replicaof`, with args beginning with: `10.10.10.10`, `6379`,
```

And do not forget to check that after the upgrade, provided that the change has been applied, the data is still there. You can try creating some rate limit in 3scale before upgrade, for instance 5 request every day. Run 2 requests. Then upgrade 3scale and check there are only 3 requests left before having 429 response.

**Comment by valerymo** [comment]
> Looks like redis is **not** being rolled out,

Yes, rechecked, and redis pods are not rolled out, although ConfigMap updated

**Comment by valerymo** [comment]
new commit added yesterday - restrictions we are adding in this PR are Only for Internal Redis

**Comment by valerymo** [comment]
Upgrade tested after commit 4:
```
$ kubectl rollout history deployment/system-redis
deployment.apps/system-redis 
REVISION  CHANGE-CAUSE
1         <none>
2         <none>


vmogilev@vmogilev THREESCALE-9308-Redis]$ kubectl exec -ti $(kubectl get pods -l deployment=backend-redis -o name) -- /bin/bash
bash-4.2$ replicaof
bash: replicaof: command not found
bash-4.2$ cd /     
bash-4.2$ find . -name "redis.conf"
./etc/opt/rh/rh-redis6/redis.conf
./etc/redis.d/..2024_03_14_15_54_49.3945022385/redis.conf
./etc/redis.d/redis.conf
./etc/redis.conf
./opt/rh/rh-redis6/register.content/etc/opt/rh/rh-redis6/redis.conf
bash-4.2$ cat ./etc/redis.d/..2024_03_14_15_54_49.3945022385/redis.conf |grep REPLICAOF
rename-command REPLICAOF ""
bash-4.2$ 
```

**Comment by valerymo** [comment]
- Will Open new PR and close this one as significant changes in master, including GO version
- Please continue review in PR https://github.com/3scale/3scale-operator/pull/954
- All tests will be repeated in new PR - 954
cc @eguzki , @carlkyrillos

**Pull request closed without merging**